###################################################################################################
Date: Fri, 7 Dec 2018 15:06:42 +0000
From 06f45b8dd4fed624f8d9afade9a5a84f0b6d7542 Mon Sep 17 00:00:00 2001
From: Ben Gray <ben.gray@sky.uk>
Subject: [PATCH] Added config option to power off the adapter when bluetoothd starts.

This is useful in crash / restart situations as it allows the bluetoothd and kernel to synchronise
their states.

Source: SKY
Upstream-Status: Pending
Signed-off-by: Tijo Thomas John <tijo.thomas@sky.uk>
###################################################################################################

---
 src/adapter.c | 8 ++++++--
 src/hcid.h    | 1 +
 src/main.c    | 9 +++++++++
 src/main.conf | 6 ++++++
 4 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/src/adapter.c b/src/adapter.c
index d6e1ac7..28619ee 100644
--- a/src/adapter.c
+++ b/src/adapter.c
@@ -8450,8 +8450,12 @@ static void read_info_complete(uint8_t status, uint16_t length,
 	if (adapter->stored_discoverable && !adapter->discoverable_timeout)
 		set_discoverable(adapter, 0x01, 0);
 
-	if (adapter->current_settings & MGMT_SETTING_POWERED)
-		adapter_start(adapter);
+	if (adapter->current_settings & MGMT_SETTING_POWERED) {
+		if (main_opts.startup_power_off)
+			set_mode(adapter, MGMT_OP_SET_POWERED, 0x00);
+		else
+			adapter_start(adapter);
+	}
 
 	return;
 
diff --git a/src/hcid.h b/src/hcid.h
index 74ad583..55ad281 100644
--- a/src/hcid.h
+++ b/src/hcid.h
@@ -57,6 +57,7 @@ struct main_opts {
 
 	gboolean	kern_conn_control;
 	gboolean	secure_conn;
+	gboolean	startup_power_off;
 };
 
 extern struct main_opts main_opts;
diff --git a/src/main.c b/src/main.c
index 6a85159..f74c50e 100644
--- a/src/main.c
+++ b/src/main.c
@@ -111,6 +111,7 @@ static const char *gatt_options[] = {
 static const char *sky_options[] = {
 	"KernelConnectionControl",
 	"SecureConnections",
+	"StartupPowerOff",
 	NULL
 };
 
@@ -319,6 +320,13 @@ static void parse_sky_config(GKeyFile *config)
 		main_opts.secure_conn = boolean;
 	}
 
+	boolean = g_key_file_get_boolean(config, "sky",
+						"StartupPowerOff", &err);
+	if (err)
+		g_clear_error(&err);
+	else
+		main_opts.startup_power_off = boolean;
+
 }
 
 static void parse_config(GKeyFile *config)
@@ -496,6 +504,7 @@ static void init_defaults(void)
 	/* Default settings for Sky extensions */
 	main_opts.kern_conn_control = TRUE;
 	main_opts.secure_conn = TRUE;
+	main_opts.startup_power_off = FALSE;
 }
 
 static void log_handler(const gchar *log_domain, GLogLevelFlags log_level,
diff --git a/src/main.conf b/src/main.conf
index 9efcd7c..8d714e8 100644
--- a/src/main.conf
+++ b/src/main.conf
@@ -113,3 +113,9 @@
 # interopt issues with some BLE devices. Defaults to 'true'.
 # SecureConnections = true
 
+# Sets whether the adapter should be powered off at start-up if already powered
+# on.  This is useful to reset the adapter and therefore have the kernel and
+# userspace in sync. Defaults to 'false'
+# StartupPowerOff = false
+
+
-- 
2.18.0

